File: hackerrank/data-structures/balanced-trees/median-updates/median-updates.java
// HackerRank - Median Updates
// https://www.hackerrank.com/challenges/median
// In Java 8, because that's what HackerRank supports. The problem is intended
// to be solved with self-balancing trees, but this solution uses binary heaps
// equipped with an O(log(n)) remove operation.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * A priority queue with an O(log(n)) remove operation.
 * <p>
 *   Implemented as a binary heap augmented with a table mapping each key to
 *   all of its occurrences in the heap.
 * </p>
 **/
final class BinaryHeap<E> {
    static <E extends Comparable<E>> BinaryHeap<E> makeMinHeap() {
        return new BinaryHeap<>(Comparator.naturalOrder());
    }

    static <E extends Comparable<E>> BinaryHeap<E> makeMaxHeap() {
        return new BinaryHeap<>(Collections.reverseOrder());
    }

    boolean isEmpty() {
        return _heap.isEmpty();
    }

    int size() {
        return _heap.size();
    }

    void clear() {
        _heap.clear();
        _map.clear();
    }

    void push(E value) {
        _heap.add(null);
        int child = siftUp(value);
        _map.computeIfAbsent(value, val -> new HashSet<>()).add(child);
        _heap.set(child, value);
    }

    E pop() {
        E value = peek();

        if (size() == 1) {
            clear();
        } else {
            _map.computeIfPresent(value, (val, indices) -> {
                if (indices.size() == 1) {
                    return null; // Remove the mapping.
                } else {
                    indices.remove(0); // Remove the index from the mapped set.
                    return indices; // Keep the mapping.
                }
            });

            cutRoot();
        }

        return value;
    }

    E peek() {
        return _heap.get(0);
    }

    boolean remove(E value) {
        Set<Integer> indices = _map.get(value);
        if (indices == null) return false;

        if (size() == 1) {
            clear();
        } else {
            // Get any index to this value in the heap.
            int child = indices.iterator().next();

            if (indices.size() == 1) {
                _map.remove(value); // Remove the mapping.
            } else {
                child = highestEqualAncestor(child);
                indices.remove(child); // Remove the index from the mapped set.
            }

            yankUp(child);
            cutRoot();
        }

        return true;
    }

    private static final int NO_CHILD = -1;

    private BinaryHeap(Comparator<? super E> comp) {
        _comp = comp;
    }

    private void cutRoot() {
        int last = size() - 1;
        E value = _heap.get(last);
        _heap.remove(last); // Remove the item whose index is last.

        Set<Integer> indices = _map.get(value);
        indices.remove(last); // Remove the item comparing equal to last.

        int index = siftDown(value);
        indices.add(index);
        _heap.set(index, value);
    }

    private int highestEqualAncestor(int child) {
        E value = _heap.get(child);

        while (child != 0) {
            int parent = (child - 1) / 2;
            if (!_heap.get(parent).equals(value)) break;
            child = parent;
        }

        return child;
    }

    private void yankUp(int child) {
        while (child != 0) {
            int parent = (child - 1) / 2;
            E parentValue = _heap.get(parent);

            Set<Integer> indices = _map.get(parentValue);
            indices.remove(parent);
            indices.add(child);
            _heap.set(child, parentValue);

            child = parent;
        }
    }

    private int siftUp(E childValue) {
        int child = size() - 1;

        while (child != 0) {
            int parent = (child - 1) / 2;
            E parentValue = _heap.get(parent);
            if (orderOk(parentValue, childValue)) break;

            Set<Integer> indices = _map.get(parentValue);
            indices.remove(parent);
            indices.add(child);
            _heap.set(child, parentValue);

            child = parent;
        }

        return child;
    }

    private int siftDown(E parentValue) {
        int parent = 0;

        for (; ; ) {
            int child = pickChild(parent);
            if (child == NO_CHILD) break;
            E childValue = _heap.get(child);
            if (orderOk(parentValue, childValue)) break;

            Set<Integer> indices = _map.get(childValue);
            indices.remove(child);
            indices.add(parent);
            _heap.set(parent, childValue);

            parent = child;
        }

        return parent;
    }

    private int pickChild(int parent) {
        int left = parent * 2 + 1;
        if (left >= size()) return NO_CHILD;

        int right = left + 1;
        return right == size() || orderOk(_heap.get(left), _heap.get(right))
                ? left
                : right;
    }

    private boolean orderOk(E parentValue, E childValue) {
        return _comp.compare(parentValue, childValue) <= 0;
    }

    private final Map<E, Set<Integer>> _map = new HashMap<>();

    private final List<E> _heap = new ArrayList<>();

    private final Comparator<? super E> _comp;
}

final class MedianBag {
    boolean isEmpty() {
        return _low.isEmpty() && _high.isEmpty();
    }

    void add(int value) {
        if (!_low.isEmpty() && value < _low.peek())
            _low.push(value);
        else
            _high.push(value);

        rebalance();
    }

    boolean remove(int value) {
        if (_low.remove(value) || _high.remove(value)) {
            rebalance();
            return true;
        }

        return false;
    }

    void printMedian() {
        switch (balanceFactor()) {
        case -1:
            System.out.println(_low.peek());
            break;

        case +1:
            System.out.println(_high.peek());
            break;

        case 0:
            long sum = (long)_low.peek() + (long)_high.peek();
            long half = sum / 2;

            if (sum % 2 == 0)
                System.out.println(half);
            else if (sum == -1)
                System.out.println("-0.5");
            else
                System.out.println(half + ".5");

            break;

        default:
            throw new AssertionError("Bug: balancing invariant violated");
        }
    }

    private void rebalance() {
        switch (balanceFactor()) {
        case -2:
            _high.push(_low.pop());
            break;

        case +2:
            _low.push(_high.pop());
            break;

        default:
            break; // Can't be made more balanced.
        }
    }

    private int balanceFactor() {
        return _high.size() - _low.size();
    }

    private final BinaryHeap<Integer> _low = BinaryHeap.makeMaxHeap();

    private final BinaryHeap<Integer> _high = BinaryHeap.makeMinHeap();
}

enum Solution {
    ;

    public static void main(String[] args) throws IOException {
        MedianBag bag = new MedianBag();

        try (   Reader isr = new InputStreamReader(System.in);
                BufferedReader br = new BufferedReader(isr)) {
            Matcher matcher = TOKEN.matcher(br.readLine());
            for (int count = nextInt(matcher); count > 0; --count) {
                matcher.reset(br.readLine());
                char opcode = next(matcher).charAt(0);
                int argument = nextInt(matcher);

                switch (opcode) {
                case 'a':
                    bag.add(argument);
                    break;

                case 'r':
                    if (!bag.remove(argument) || bag.isEmpty()) {
                        System.out.println("Wrong!");
                        continue;
                    }
                    break;

                default:
                    break; // Ignore unrecognized opcodes.
                }

                bag.printMedian();
            }
        }
    }

    private static String next(Matcher matcher) {
        matcher.find();
        return matcher.group();
    }

    private static int nextInt(Matcher matcher) {
        return Integer.parseInt(next(matcher));
    }

    private static final Pattern TOKEN = Pattern.compile("\\S+");
}

File: hackerrank/data-structures/balanced-trees/median-updates/median-updates.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# HackerRank - Median Updates
# https://www.hackerrank.com/challenges/median
# In Crystal. The problem is intended to be solved with self-balancing trees,
# but this solution uses binary heaps equipped with an O(log(n)) delete
# operation.

$VERBOSE = 1

require 'set'

# A priority queue with an O(log(n)) remove operation, implemented as a binary
# heap augmented with a table mapping each key to all of its occurrences in the
# heap.
class BinaryHeap
  def self.make_min_heap
    BinaryHeap.new { |ls, rs| ls <=> rs }
  end

  def self.make_max_heap
    BinaryHeap.new { |ls, rs| rs <=> ls }
  end

  def empty?
    @heap.empty?
  end

  def size
    @heap.size
  end

  def clear
    @map.clear
    @heap.clear
  end

  def push(value)
    child = sift_up(value)
    (@map[value] ||= Set.new).add(child)
    @heap[child] = value
    nil
  end

  def pop
    value = first

    if size == 1
      clear
    else
      indices = @map[value]

      if indices.size == 1
        @map.delete(value)
      else
        indices.delete(0)
      end

      cut_root
    end

    value
  end

  def first
    @heap.first
  end

  def delete?(value)
    indices = @map[value]
    return false if indices.nil?

    if size == 1
      clear
    else
      child = indices.first # Get any index to this value in the heap.

      if indices.size == 1
        @map.delete(value)
      else
        child = highest_equal_ancestor(child)
        indices.delete(child)
      end

      yank_up(child)
      cut_root
    end

    true
  end

  protected

  def initialize(&comparer)
    @map = {}  # value => indices
    @heap = [] # index => value
    @comparer = comparer
  end

  private

  def cut_root
    value = @heap.pop
    indices = @map[value]
    indices.delete(size)

    index = sift_down(value)
    indices.add(index)
    @heap[index] = value

    nil
  end

  def highest_equal_ancestor(child)
    value = @heap[child]

    until child.zero?
      parent = (child - 1) / 2
      break if @heap[parent] != value

      child = parent
    end

    child
  end

  def yank_up(child)
    until child.zero?
      parent = (child - 1) / 2
      parent_value = @heap[parent]

      @map[parent_value].delete(parent).add(child)
      @heap[child] = parent_value

      child = parent
    end
    nil
  end

  def sift_up(child_value)
    child = size

    until child.zero?
      parent = (child - 1) / 2
      parent_value = @heap[parent]
      break if order_ok?(parent_value, child_value)

      @map[parent_value].delete(parent).add(child)
      @heap[child] = parent_value

      child = parent
    end

    child
  end

  def sift_down(parent_value)
    parent = 0

    loop do
      child = pick_child(parent)
      break if child.nil?

      child_value = @heap[child]
      break if order_ok?(parent_value, child_value)

      @map[child_value].delete(child).add(parent)
      @heap[parent] = child_value

      parent = child
    end

    parent
  end

  def pick_child(parent)
    left = parent * 2 + 1
    return nil if left >= size

    right = left + 1
    right == size || order_ok?(@heap[left], @heap[right]) ? left : right
  end

  def order_ok?(parent_value, child_value)
    @comparer.call(parent_value, child_value) <= 0
  end
end

# A multiset with O(log(n)) insertion and deletion and O(1) median finding.
class MedianBag
  def initialize
    @low = BinaryHeap.make_max_heap
    @high = BinaryHeap.make_min_heap
  end

  def empty?
    @low.empty? && @high.empty?
  end

  def add(value)
    if !@low.empty? && value < @low.first
      @low.push(value)
    else
      @high.push(value)
    end

    rebalance
  end

  def delete?(value)
    if @low.delete?(value) || @high.delete?(value)
      rebalance
      true
    else
      false
    end
  end

  def median
    case balance_factor
    when -1
      @low.first
    when +1
      @high.first
    when 0
      (@low.first + @high.first) / 2.0
    else
      raise "Bug: balancing invariant violated"
    end
  end

  private

  def rebalance
    case balance_factor
    when -2
      @high.push(@low.pop)
    when +2
      @low.push(@high.pop)
    end # In other cases, it can't be made more balanced.

    nil
  end

  def balance_factor
    @high.size - @low.size
  end
end

def print_without_trailing_fractional_zeros(value)
  if value.is_a?(Float) && (rounded_value = value.to_i) == value
    puts rounded_value
  else
    puts value
  end
end

if __FILE__ == $PROGRAM_NAME
  bag = MedianBag.new

  gets.to_i.times do
    tokens = gets.split
    opcode = tokens[0]
    argument = tokens[1].to_i

    case opcode
    when 'a'
      bag.add(argument)
    when 'r'
      if !bag.delete?(argument) || bag.empty?
        puts 'Wrong!'
        next
      end
    else
      raise 'Unrecognized opcode.'
    end

    print_without_trailing_fractional_zeros(bag.median)
  end
end
